<a href="https://colab.research.google.com/github/chibaba/DATA-SCIENCE-AND-MACHINE-LEARNING/blob/master/Feature_Engineering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [12]:
import pandas as pd


In [13]:
file_url = 'https://raw.githubusercontent.com/PacktWorkshops/The-Data-Science-Workshop/master/Chapter03/bank-full.csv'

In [14]:
# Reading the banking data
bankData = pd.read_csv(file_url, sep=";")

In [15]:
bankData.groupby(['housing', 'y'])['y'].agg(houseTot='count').reset_index()


housing    y  houseTot
0      no   no     16727
1      no  yes      3354
2     yes   no     23195
3     yes  yes      1935

The first part of the code is to group customers based on whether they own a house or not. The count of customers under each category is calculated with the .agg() method. From the values, we can see that the propensity to buy term deposits is much higher for people who do not own a house compared with those who do own one: ( 3354 / ( 3354 + 16727) = 17% to 1935 / ( 1935 + 23195) = 8%).



In [16]:
# Relationship between having a loan and propensity for term deposits
bankData.groupby(['loan', 'y'])['y'].agg(loanTot='count').reset_index()

loan    y  loanTot
0   no   no    33162
1   no  yes     4805
2  yes   no     6760
3  yes  yes      484

In [17]:
#Taking the quantiles for 25%, 50% and 75% of the balance data
import numpy as np
np.quantile(bankData['balance'],[0.25,0.5,0.75])

array([  72.,  448., 1428.])

In [18]:
bankData['balanceClass'] = 'Quant1'
bankData.loc[(bankData['balance'] > 72) & (bankData['balance'] < 448), 'balanceClass'] = 'Quant2'
bankData.loc[(bankData['balance'] > 448) & (bankData['balance'] < 1428), 'balanceClass'] = 'Quant3'
bankData.loc[bankData['balance'] > 1428, 'balanceClass'] = 'Quant4'
bankData.head()

age           job  marital  education  ... previous  poutcome   y balanceClass
0   58    management  married   tertiary  ...        0   unknown  no       Quant4
1   44    technician   single  secondary  ...        0   unknown  no       Quant1
2   33  entrepreneur  married  secondary  ...        0   unknown  no       Quant1
3   47   blue-collar  married    unknown  ...        0   unknown  no       Quant4
4   33       unknown   single    unknown  ...        0   unknown  no       Quant1

[5 rows x 18 columns]

In [19]:
# Calculating the total customers categorised as per quantile and propensity classification
balanceProp = bankData.groupby(['balanceClass', 'y'])['y'].agg(balanceCat='count').reset_index()
balanceProp

balanceClass    y  balanceCat
0       Quant1   no       10517
1       Quant1  yes         823
2       Quant2   no       10049
3       Quant2  yes        1226
4       Quant3   no        9884
5       Quant3  yes        1415
6       Quant4   no        9472
7       Quant4  yes        1825